new_installer.py: docs (#52164)
LLNL/spack: lib/spack/docs/installing.rst

This is the primary concurrency knob.

.. code-block:: console

   $ spack install -j16 python

Spack creates a POSIX jobserver compatible with GNU Make's jobserver protocol.
Child build systems automatically respect it through ``MAKEFLAGS``, so total CPU usage stays bounded regardless of how many packages are building concurrently.

.. note::

   If an external jobserver is already present in ``MAKEFLAGS``, for example when Spack itself is invoked from inside a larger ``make`` build, Spack attaches to the existing jobserver instead of creating its own.

Package-level parallelism (``-p``)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The ``-p`` / ``--concurrent-packages`` flag limits how many packages can be in the build queue simultaneously.
By default there is no limit, and packages are started as jobserver tokens become available.

.. code-block:: console

   $ spack install -j16 -p4 python

This builds with 16 total make-jobs but never more than 4 packages at once.

Dynamic adjustment
^^^^^^^^^^^^^^^^^^

You can adjust parallelism while a build is running:

* Press ``+`` to add a job (increases ``-j`` by 1)
* Press ``-`` to remove a job (decreases ``-j`` by 1)

> The progress header shows the adjustment in progress, e.g. ``+/-: 4=>2 jobs``, until the actual count reaches the target.

Reviewer: ```suggestion
No jobs are killed to adjust the parallelism.
There is a delay between reducing the number of jobs and sufficient jobs completing for the jobserver size to be reduced.
The progress header shows the adjustment in progress, e.g. ``+/-: 4=>2 jobs``, until the actual count reaches the target.
```

Author: Good note to add, I rephrased it a little bit

Multi-process and multi-node installs
--------------------------------------

Multiple ``spack install`` processes can run concurrently against the same store, which is useful on clusters where different nodes install different parts of a software stack.
Spack coordinates through per-prefix filesystem locks: before building a package, the process acquires an exclusive lock on its install prefix.
If another process already holds the lock, Spack waits rather than building a second copy.
When a process encounters a prefix that was already installed, it simply skips it and moves on to the next dependency.

Non-interactive mode
--------------------

> In CI pipelines, when redirecting output to a file, or when running in the background, Spack skips the TUI and prints simple line-based status updates instead.

Reviewer: ```suggestion
When the controlling process is not a tty, such as In CI pipelines, when redirecting output to a file, or when running in the background, Spack skips the TUI and prints simple line-based status updates instead.
```

> Use ``-v`` to also print build output.

Reviewer: ```suggestion
Use ``spack install -v`` to also print build output.
```

You can also background builds:

* **Suspend and resume**: press ``Ctrl-Z`` to suspend the install, then ``bg`` to let it continue in the background or ``fg`` to bring it back.

>   Child builds are paused while suspended and resumed when foregrounded.

Reviewer: ```suggestion
  Child builds are paused while suspended, and resumed when continued in the background or foreground.
```
